syntax = "proto3";
import "google/protobuf/timestamp.proto";


package distlock;


// A distributed lock service
service Distlock {
  // Creates a new lock with the given key on the server.
  rpc CreateLock(Lock) returns (EmptyResponse) {}

  // Acquires the lock with the given key from the server.
  rpc AcquireLock(AcquireLockRequest) returns (Lock) {}

  // Releases the lock with the given key from the server.
  rpc ReleaseLock(Lock) returns (EmptyResponse) {}
}


// The empty response.
message EmptyResponse {}


// The message for a lock object.
message Lock {
  string key = 1;
  bool acquired = 2;
  int64 clock = 3;
  google.protobuf.Timestamp timeout = 4;
}


// The request message containing the key name of the lock to acquire.
message AcquireLockRequest {
  string key = 1;
  int64 timeout_seconds = 2;
}
